source/extensions/geoip_providers/maxmind/config.cc

```diff
@@ -17,58 +17,55 @@ using ConfigProto = envoy::extensions::geoip_providers::maxmind::v3::MaxMindConf
- * driver/provider. When given different configs, the singleton returns different provider
- * instances.
+ * When given equivalent provider configs, the singleton returns pointers to the same driver/provider.
+ * When given different configs, the singleton returns different provider instances.
  */
 class DriverSingleton : public Envoy::Singleton::Instance {
-public:
-  std::shared_ptr<GeoipProvider> get(std::shared_ptr<DriverSingleton> singleton,
-                                     const ConfigProto& proto_config,
-                                     const std::string& stat_prefix,
-                                     Server::Configuration::FactoryContext& context) {
-    std::shared_ptr<GeoipProvider> driver;
-    const uint64_t key = MessageUtil::hash(proto_config);
-    absl::MutexLock lock(&mu_);
-    auto it = drivers_.find(key);
-    if (it != drivers_.end()) {
-      driver = it->second.lock();
-    } else {
-      const auto& provider_config =
-          std::make_shared<GeoipProviderConfig>(proto_config, stat_prefix, context.scope());
-      driver = std::make_shared<GeoipProvider>(singleton, provider_config);
-      drivers_[key] = driver;
-    }
-    return driver;
+  public:
+    std::shared_ptr<GeoipProvider> get(std::shared_ptr<DriverSingleton> singleton,
+                                           const ConfigProto& proto_config,
+                                           const std::string& stat_prefix, Server::Configuration::FactoryContext& context) {
+      std::shared_ptr<GeoipProvider> driver;
+      const uint64_t key = MessageUtil::hash(proto_config);
+      absl::MutexLock lock(&mu_);
+      auto it = drivers_.find(key);
+      if (it != drivers_.end()) {
+        driver = it->second.lock();
+      } else {
+        const auto& provider_config =
+        std::make_shared<GeoipProviderConfig>(proto_config, stat_prefix, context.scope());
+        driver = std::make_shared<GeoipProvider>(singleton, provider_config);
+        drivers_[key] = driver;
+      }
+      return driver;
   }
-
-private:
-  absl::Mutex mu_;
-  // We keep weak_ptr here so the providers can be destroyed if the config is updated to stop using
-  // that config of the provider. Each provider stores shared_ptrs to this singleton, which keeps
-  // the singleton from being destroyed unless it's no longer keeping track of any providers. (The
-  // singleton shared_ptr is *only* held by cache instances.)
-  absl::flat_hash_map<uint64_t, std::weak_ptr<GeoipProvider>> drivers_ ABSL_GUARDED_BY(mu_);
+  private:
+    absl::Mutex mu_;
+    // We keep weak_ptr here so the providers can be destroyed if the config is updated to stop using
+    // that config of the provider. Each provider stores shared_ptrs to this singleton, which keeps the
+    // singleton from being destroyed unless it's no longer keeping track of any providers.
+    // (The singleton shared_ptr is *only* held by cache instances.)
```

Comment:
This is not cache instances, it's drivers.

Reply:
Sorry, I was wrong, it's `GeoIpProvider` instances that keep the singleton alive, not drivers. (I just checked because I wasn't sure they were actually being kept alive at all!)

Reply:
no worries, i have used "driver" and "provider" interchangeably in the code, that may be confusing